include/zephyr/bluetooth/ead.h

```diff
@@ -0,0 +1,73 @@
+/* Copyright (c) 2023 Nordic Semiconductor ASA
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stddef.h>
+#include <stdint.h>
+
+#include <zephyr/kernel.h>
+
+/** Additional Authenticated Data size in bytes */
+#define BT_EAD_AAD_SIZE	       1
+/** Randomizer size in bytes */
+#define BT_EAD_RANDOMIZER_SIZE 5
+/** nonce size in bytes */
+#define BT_EAD_NONCE_SIZE      13
+/** Key size in bytes */
+#define BT_EAD_KEY_SIZE	       16
+/** Initialisation Vector size in bytes */
+#define BT_EAD_IV_SIZE	       8
+/** MIC size in bytes */
+#define BT_EAD_MIC_SIZE	       4
+
+/**
+ * @brief Encrypt and authenticate the given advertising data. The resulting
+ * data in @p encrypted_ad_data will look like that:
+ *
+ *     - Randomizer is added in the 5 first bytes;
+ *     - AAD is added in the next byte;
+ *     - Encrypted payload is added (payload_size bytes);
+ *     - MIC is added in the last 4 bytes.
+ *
+ * @note The function must be called each time the RPA is updated or the data
+ * are modified.
+ *
+ * @note The term `advertising structure` is used to describe the advertising
+ *       data with the advertising type and the length of those two.
+ *
+ * @param[in]  session_key Key of 16 bytes used for the encryption.
+ * @param[in]  iv Initialisation Vector used to generate the nonce. It must be
+ *             changed each time the Session Key change.
+ * @param[in]  payload Advertising Data to encrypt. Can be multiple advertising
+ *             structure concatenated.
+ * @param[in]  payload_size Size of the Advertising Data to encrypt.
+ * @param[out] encrypted_ad_data Encrypted Adv Data including the Randomizer and
+ *             the MIC. Size must be `BT_EAD_RANDOMIZER_SIZE + BT_EAD_AAD_SIZE +
+ *             payload_size + BT_EAD_MIC_SIZE`.
+ * @retval int Size of @p encrypted_ad_data in bytes if successful, else return
+ *             `-EIO` if an error occurred during the encryption.
+ */
+int bt_ead_encrypt(const uint8_t *session_key, const uint8_t *iv, const uint8_t *payload,
+		   int payload_size, uint8_t *encrypted_ad_data);
```

Comment:
Actually we can't because `payload` can be multiple advertising structure concatenated together.
Using it for `encrypted_ad_data` could work but I in my opinion it's not the role of this function to create the advertising structure.

Reply:
> because the data field of the bt_data structure is const *uint8_t. It mean that the user should still provide an adequate buffer.

Yes, you wouldn't be able to do in-line encryption, but that isn't possible anyways since you have `const uint8_t *payload`. Not sure why using `bt_data` is different here (for the input), but yes, we cannot use `bt_data` for the output.

Reply:
Yes, using `bt_data` for encryption is not a problem at all. But indeed, for the decryption, we can't output a `bt_data`. That create asymmetrical functions, I think that it can be confusing when you use the library.

Reply:
That is a good point. Given the `constant` constraint, it may be impossible to use `bt_data` for this, if we want symmetrical functions. 

Since the output data is also LTV encoded, perhaps the output should be a `struct net_buf_simple *buf` so that it matches the data structure of the scanning `recv` callback, and makes it possible to forward to `bt_data_parse` to parse the decrypted data?

Reply:
> Since the output data is also LTV encoded, perhaps the output should be a `struct net_buf_simple *buf` so that it matches the data structure of the scanning `recv` callback, and makes it possible to forward to `bt_data_parse` to parse the decrypted data?

I tested that option but it's a bit irrelevant for `bt_ead_decrypt` to take a `net_buf_simple`. If you look at the test (https://github.com/zephyrproject-rtos/zephyr/pull/55225/files#diff-c8109c56879494f942802cad9eaa6352c80d4a4b5f61010de723adfa176665deR32), to decrypt the data I create a `net_buf_simple` but I only send the `data` field of the `net_buf_simple` to `bt_ead_decrypt`. And then I can easily send the `net_buf_simple` to `bt_data_parse`.